NullPointerException when using client scope policy for token-exchange

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

authorization-services

### Describe the bug

[RFC 8693](https://datatracker.ietf.org/doc/html/rfc8693) security section suggests hardening of token exchange flows using scopes:

> In addition, both delegation and impersonation introduce unique security issues. Any time one principal is delegated the rights of another principal, the potential for abuse is a concern. The use of the scope claim (in addition to other typical constraints such as a limited token lifetime) is suggested to mitigate potential for such abuse, as it restricts the contexts in which the delegated rights can be exercised.

So I've tried setting it up using `source` and `target` client ids.
1. Created a new client scope named `token-exchange` and added a hardcoded claim mapper that adds it to the access token as `"token-exchange":true`.
2. Created a _client scope_ policy in the `realm-management` client, which requires a `token-exchange` scope
3. Added the scope to the `source` client scopes
4.  Created a permission on the `target` client to allow token-exchange with the policy from step 2.

Then, I request a new access token from the `source` client:
```bash
curl --location 'https://staging/realms/master/protocol/openid-connect/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'client_id=source' \
--data-urlencode 'client_secret=[REDACTED]' \
--data-urlencode 'grant_type=client_credentials' \
--data-urlencode 'scope=token-exchange'
```
And when I attempt to exchange it I get `500 Internal Server Error`.
```bash
curl --location 'https://staging/realms/master/protocol/openid-connect/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'client_id=source' \
--data-urlencode 'client_secret=[REDACTED]' \
--data-urlencode 'subject_token=TOKEN' \
--data-urlencode 'audience=target' \
--data-urlencode 'grant_type=urn:ietf:params:oauth:grant-type:token-exchange' \
--data-urlencode 'requested_token_type=urn:ietf:params:oauth:token-type:access_token'
```


### Version

22.0.3

### Expected behavior

Client scope policy works

### Actual behavior

500 Internal Error
```json
{
  "refId": 1,
  "exceptionType": "java.lang.RuntimeException",
  "message": "Failed to evaluate permissions",
  "frames": [
    {
      "class": "org.keycloak.authorization.policy.evaluation.DecisionPermissionCollector",
      "method": "onError",
      "line": 182
    },
    {
      "class": "org.keycloak.authorization.permission.evaluator.IterablePermissionEvaluator",
      "method": "evaluate",
      "line": 71
    },
    {
      "class": "org.keycloak.authorization.permission.evaluator.IterablePermissionEvaluator",
      "method": "evaluate",
      "line": 87
    },
    {
      "class": "org.keycloak.services.resources.admin.permissions.MgmtPermissions",
      "method": "evaluatePermission",
      "line": 349
    },
    {
      "class": "org.keycloak.services.resources.admin.permissions.MgmtPermissions",
      "method": "evaluatePermission",
      "line": 342
    },
    {
      "class": "org.keycloak.services.resources.admin.permissions.ClientPermissions",
      "method": "canExchangeTo",
      "line": 351
    },
    {
      "class": "org.keycloak.protocol.oidc.DefaultTokenExchangeProvider",
      "method": "exchangeClientToClient",
      "line": 330
    },
    {
      "class": "org.keycloak.protocol.oidc.DefaultTokenExchangeProvider",
      "method": "tokenExchange",
      "line": 240
    },
    {
      "class": "org.keycloak.protocol.oidc.DefaultTokenExchangeProvider",
      "method": "exchange",
      "line": 126
    },
    {
      "class": "org.keycloak.protocol.oidc.endpoints.TokenEndpoint",
      "method": "tokenExchange",
      "line": 856
    },
    {
      "class": "org.keycloak.protocol.oidc.endpoints.TokenEndpoint",
      "method": "processGrantRequestInternal",
      "line": 231
    },
    {
      "class": "org.keycloak.protocol.oidc.endpoints.TokenEndpoint",
      "method": "processGrantRequest",
      "line": 190
    },
    {
      "class": "jdk.internal.reflect.GeneratedMethodAccessor67",
      "method": "invoke"
    },
    {
      "class": "jdk.internal.reflect.DelegatingMethodAccessorImpl",
      "method": "invoke",
      "line": 43
    },
    {
      "class": "java.lang.reflect.Method",
      "method": "invoke",
      "line": 568
    },
    {
      "class": "org.jboss.resteasy.core.MethodInjectorImpl",
      "method": "invoke",
      "line": 154
    },
    {
      "class": "org.jboss.resteasy.core.MethodInjectorImpl",
      "method": "invoke",
      "line": 118
    },
    {
      "class": "org.jboss.resteasy.core.ResourceMethodInvoker",
      "method": "internalInvokeOnTarget",
      "line": 560
    },
    {
      "class": "org.jboss.resteasy.core.ResourceMethodInvoker",
      "method": "invokeOnTargetAfterFilter",
      "line": 452
    },
    {
      "class": "org.jboss.resteasy.core.ResourceMethodInvoker",
      "method": "lambda$invokeOnTarget$2",
      "line": 413
    },
    {
      "class": "org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext",
      "method": "filter",
      "line": 321
    },
    {
      "class": "org.jboss.resteasy.core.ResourceMethodInvoker",
      "method": "invokeOnTarget",
      "line": 415
    },
    {
      "class": "org.jboss.resteasy.core.ResourceMethodInvoker",
      "method": "invoke",
      "line": 378
    },
    {
      "class": "org.jboss.resteasy.core.ResourceLocatorInvoker",
      "method": "invokeOnTargetObject",
      "line": 174
    },
    {
      "class": "org.jboss.resteasy.core.ResourceLocatorInvoker",
      "method": "invoke",
      "line": 142
    },
    {
      "class": "org.jboss.resteasy.core.ResourceLocatorInvoker",
      "method": "invokeOnTargetObject",
      "line": 168
    },
    {
      "class": "org.jboss.resteasy.core.ResourceLocatorInvoker",
      "method": "invoke",
      "line": 131
    },
    {
      "class": "org.jboss.resteasy.core.ResourceLocatorInvoker",
      "method": "invoke",
      "line": 33
    },
    {
      "class": "org.jboss.resteasy.core.SynchronousDispatcher",
      "method": "invoke",
      "line": 429
    },
    {
      "class": "org.jboss.resteasy.core.SynchronousDispatcher",
      "method": "lambda$invoke$4",
      "line": 240
    },
    {
      "class": "org.jboss.resteasy.core.SynchronousDispatcher",
      "method": "lambda$preprocess$0",
      "line": 154
    },
    {
      "class": "org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext",
      "method": "filter",
      "line": 321
    },
    {
      "class": "org.jboss.resteasy.core.SynchronousDispatcher",
      "method": "preprocess",
      "line": 157
    },
    {
      "class": "org.jboss.resteasy.core.SynchronousDispatcher",
      "method": "invoke",
      "line": 229
    },
    {
      "class": "io.quarkus.resteasy.runtime.standalone.RequestDispatcher",
      "method": "service",
      "line": 82
    },
    {
      "class": "io.quarkus.resteasy.runtime.standalone.VertxRequestHandler",
      "method": "dispatch",
      "line": 147
    },
    {
      "class": "io.quarkus.resteasy.runtime.standalone.VertxRequestHandler",
      "method": "handle",
      "line": 84
    },
    {
      "class": "io.quarkus.resteasy.runtime.standalone.VertxRequestHandler",
      "method": "handle",
      "line": 44
    },
    {
      "class": "io.opentelemetry.javaagent.instrumentation.vertx.RoutingContextHandlerWrapper",
      "method": "handle",
      "line": 39
    },
    {
      "class": "io.opentelemetry.javaagent.instrumentation.vertx.RoutingContextHandlerWrapper",
      "method": "handle",
      "line": 21
    },
    {
      "class": "io.vertx.ext.web.impl.RouteState",
      "method": "handleContext",
      "line": 1284
    },
    {
      "class": "io.vertx.ext.web.impl.RoutingContextImplBase",
      "method": "iterateNext",
      "line": 177
    },
    {
      "class": "io.vertx.ext.web.impl.RoutingContextImpl",
      "method": "next",
      "line": 141
    },
    {
      "class": "io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$1",
      "method": "handle",
      "line": 58
    },
    {
      "class": "io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$1",
      "method": "handle",
      "line": 36
    },
    {
      "class": "io.opentelemetry.javaagent.instrumentation.vertx.RoutingContextHandlerWrapper",
      "method": "handle",
      "line": 39
    },
    {
      "class": "io.opentelemetry.javaagent.instrumentation.vertx.RoutingContextHandlerWrapper",
      "method": "handle",
      "line": 21
    },
    {
      "class": "io.vertx.ext.web.impl.RouteState",
      "method": "handleContext",
      "line": 1284
    },
    {
      "class": "io.vertx.ext.web.impl.RoutingContextImplBase",
      "method": "iterateNext",
      "line": 177
    },
    {
      "class": "io.vertx.ext.web.impl.RoutingContextImpl",
      "method": "next",
      "line": 141
    },
    {
      "class": "org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter",
      "method": "lambda$createBlockingHandler$0",
      "line": 82
    },
    {
      "class": "io.opentelemetry.javaagent.bootstrap.executors.ContextPropagatingRunnable",
      "method": "run",
      "line": 37
    },
    {
      "class": "io.quarkus.vertx.core.runtime.VertxCoreRecorder$14",
      "method": "runWith",
      "line": 576
    },
    {
      "class": "org.jboss.threads.EnhancedQueueExecutor$Task",
      "method": "run",
      "line": 2513
    },
    {
      "class": "org.jboss.threads.EnhancedQueueExecutor$ThreadBody",
      "method": "run",
      "line": 1538
    },
    {
      "class": "org.jboss.threads.DelegatingRunnable",
      "method": "run",
      "line": 29
    },
    {
      "class": "org.jboss.threads.ThreadLocalResettingRunnable",
      "method": "run",
      "line": 29
    },
    {
      "class": "io.netty.util.concurrent.FastThreadLocalRunnable",
      "method": "run",
      "line": 30
    },
    {
      "class": "java.lang.Thread",
      "method": "run",
      "line": 833
    }
  ],
  "causedBy": {
    "exception": {
      "refId": 2,
      "exceptionType": "java.lang.NullPointerException",
      "message": "Cannot invoke \"org.keycloak.authorization.attribute.Attributes$Entry.asString(int)\" because the return value of \"org.keycloak.authorization.attribute.Attributes.getValue(String)\" is null",
      "frames": [
        {
          "class": "org.keycloak.authorization.policy.provider.clientscope.ClientScopePolicyProvider",
          "method": "hasClientScope",
          "line": 75
        },
        {
          "class": "org.keycloak.authorization.policy.provider.clientscope.ClientScopePolicyProvider",
          "method": "evaluate",
          "line": 61
        },
        {
          "class": "org.keycloak.authorization.policy.provider.permission.AbstractPermissionProvider",
          "method": "evaluate",
          "line": 56
        },
        {
          "class": "org.keycloak.authorization.policy.provider.permission.ScopePolicyProvider",
          "method": "evaluate",
          "line": 52
        },
        {
          "class": "org.keycloak.authorization.policy.evaluation.DefaultPolicyEvaluator",
          "method": "lambda$createPolicyEvaluator$0",
          "line": 117
        },
        {
          "class": "java.util.function.Consumer",
          "method": "lambda$andThen$0",
          "line": 65
        },
        {
          "class": "java.util.function.Consumer",
          "method": "lambda$andThen$0",
          "line": 65
        },
        {
          "class": "org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession$PolicyCache$1",
          "method": "accept",
          "line": 967
        },
        {
          "class": "org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession$PolicyCache$1",
          "method": "accept",
          "line": 962
        },
        {
          "class": "java.util.stream.ForEachOps$ForEachOp$OfRef",
          "method": "accept",
          "line": 183
        },
        {
          "class": "java.util.stream.ReferencePipeline$2$1",
          "method": "accept",
          "line": 179
        },
        {
          "class": "java.util.stream.ReferencePipeline$3$1",
          "method": "accept",
          "line": 197
        },
        {
          "class": "java.util.Iterator",
          "method": "forEachRemaining",
          "line": 133
        },
        {
          "class": "java.util.Spliterators$IteratorSpliterator",
          "method": "forEachRemaining",
          "line": 1845
        },
        {
          "class": "java.util.stream.AbstractPipeline",
          "method": "copyInto",
          "line": 509
        },
        {
          "class": "java.util.stream.AbstractPipeline",
          "method": "wrapAndCopyInto",
          "line": 499
        },
        {
          "class": "java.util.stream.ForEachOps$ForEachOp",
          "method": "evaluateSequential",
          "line": 150
        },
        {
          "class": "java.util.stream.ForEachOps$ForEachOp$OfRef",
          "method": "evaluateSequential",
          "line": 173
        },
        {
          "class": "java.util.stream.AbstractPipeline",
          "method": "evaluate",
          "line": 234
        },
        {
          "class": "java.util.stream.ReferencePipeline",
          "method": "forEach",
          "line": 596
        },
        {
          "class": "org.keycloak.utils.ClosingStream",
          "method": "forEach",
          "line": 128
        },
        {
          "class": "org.keycloak.authorization.jpa.store.JPAPolicyStore",
          "method": "findByResource",
          "line": 224
        },
        {
          "class": "org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession$PolicyCache",
          "method": "lambda$findByResource$4",
          "line": 962
        },
        {
          "class": "org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession$PolicyCache",
          "method": "cacheQuery",
          "line": 1075
        },
        {
          "class": "org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession$PolicyCache",
          "method": "cacheQuery",
          "line": 1064
        },
        {
          "class": "org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession$PolicyCache",
          "method": "findByResource",
          "line": 960
        },
        {
          "class": "org.keycloak.authorization.AuthorizationProvider$3",
          "method": "findByResource",
          "line": 409
        },
        {
          "class": "org.keycloak.authorization.policy.evaluation.DefaultPolicyEvaluator",
          "method": "evaluate",
          "line": 71
        },
        {
          "class": "org.keycloak.authorization.permission.evaluator.IterablePermissionEvaluator",
          "method": "evaluate",
          "line": 66
        },
        {
          "class": "org.keycloak.authorization.permission.evaluator.IterablePermissionEvaluator",
          "method": "evaluate",
          "line": 87
        },
        {
          "class": "org.keycloak.services.resources.admin.permissions.MgmtPermissions",
          "method": "evaluatePermission",
          "line": 349
        },
        {
          "class": "org.keycloak.services.resources.admin.permissions.MgmtPermissions",
          "method": "evaluatePermission",
          "line": 342
        },
        {
          "class": "org.keycloak.services.resources.admin.permissions.ClientPermissions",
          "method": "canExchangeTo",
          "line": 351
        },
        {
          "class": "org.keycloak.protocol.oidc.DefaultTokenExchangeProvider",
          "method": "exchangeClientToClient",
          "line": 330
        },
        {
          "class": "org.keycloak.protocol.oidc.DefaultTokenExchangeProvider",
          "method": "tokenExchange",
          "line": 240
        },
        {
          "class": "org.keycloak.protocol.oidc.DefaultTokenExchangeProvider",
          "method": "exchange",
          "line": 126
        },
        {
          "class": "org.keycloak.protocol.oidc.endpoints.TokenEndpoint",
          "method": "tokenExchange",
          "line": 856
        },
        {
          "class": "org.keycloak.protocol.oidc.endpoints.TokenEndpoint",
          "method": "processGrantRequestInternal",
          "line": 231
        },
        {
          "class": "org.keycloak.protocol.oidc.endpoints.TokenEndpoint",
          "method": "processGrantRequest",
          "line": 190
        },
        {
          "class": "jdk.internal.reflect.GeneratedMethodAccessor67",
          "method": "invoke"
        },
        {
          "class": "jdk.internal.reflect.DelegatingMethodAccessorImpl",
          "method": "invoke",
          "line": 43
        },
        {
          "class": "java.lang.reflect.Method",
          "method": "invoke",
          "line": 568
        },
        {
          "class": "org.jboss.resteasy.core.MethodInjectorImpl",
          "method": "invoke",
          "line": 154
        },
        {
          "class": "org.jboss.resteasy.core.MethodInjectorImpl",
          "method": "invoke",
          "line": 118
        },
        {
          "class": "org.jboss.resteasy.core.ResourceMethodInvoker",
          "method": "internalInvokeOnTarget",
          "line": 560
        },
        {
          "class": "org.jboss.resteasy.core.ResourceMethodInvoker",
          "method": "invokeOnTargetAfterFilter",
          "line": 452
        },
        {
          "class": "org.jboss.resteasy.core.ResourceMethodInvoker",
          "method": "lambda$invokeOnTarget$2",
          "line": 413
        },
        {
          "class": "org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext",
          "method": "filter",
          "line": 321
        },
        {
          "class": "org.jboss.resteasy.core.ResourceMethodInvoker",
          "method": "invokeOnTarget",
          "line": 415
        },
        {
          "class": "org.jboss.resteasy.core.ResourceMethodInvoker",
          "method": "invoke",
          "line": 378
        },
        {
          "class": "org.jboss.resteasy.core.ResourceLocatorInvoker",
          "method": "invokeOnTargetObject",
          "line": 174
        },
        {
          "class": "org.jboss.resteasy.core.ResourceLocatorInvoker",
          "method": "invoke",
          "line": 142
        },
        {
          "class": "org.jboss.resteasy.core.ResourceLocatorInvoker",
          "method": "invokeOnTargetObject",
          "line": 168
        },
        {
          "class": "org.jboss.resteasy.core.ResourceLocatorInvoker",
          "method": "invoke",
          "line": 131
        },
        {
          "class": "org.jboss.resteasy.core.ResourceLocatorInvoker",
          "method": "invoke",
          "line": 33
        },
        {
          "class": "org.jboss.resteasy.core.SynchronousDispatcher",
          "method": "invoke",
          "line": 429
        },
        {
          "class": "org.jboss.resteasy.core.SynchronousDispatcher",
          "method": "lambda$invoke$4",
          "line": 240
        },
        {
          "class": "org.jboss.resteasy.core.SynchronousDispatcher",
          "method": "lambda$preprocess$0",
          "line": 154
        },
        {
          "class": "org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext",
          "method": "filter",
          "line": 321
        },
        {
          "class": "org.jboss.resteasy.core.SynchronousDispatcher",
          "method": "preprocess",
          "line": 157
        },
        {
          "class": "org.jboss.resteasy.core.SynchronousDispatcher",
          "method": "invoke",
          "line": 229
        },
        {
          "class": "io.quarkus.resteasy.runtime.standalone.RequestDispatcher",
          "method": "service",
          "line": 82
        },
        {
          "class": "io.quarkus.resteasy.runtime.standalone.VertxRequestHandler",
          "method": "dispatch",
          "line": 147
        },
        {
          "class": "io.quarkus.resteasy.runtime.standalone.VertxRequestHandler",
          "method": "handle",
          "line": 84
        },
        {
          "class": "io.quarkus.resteasy.runtime.standalone.VertxRequestHandler",
          "method": "handle",
          "line": 44
        },
        {
          "class": "io.opentelemetry.javaagent.instrumentation.vertx.RoutingContextHandlerWrapper",
          "method": "handle",
          "line": 39
        },
        {
          "class": "io.opentelemetry.javaagent.instrumentation.vertx.RoutingContextHandlerWrapper",
          "method": "handle",
          "line": 21
        },
        {
          "class": "io.vertx.ext.web.impl.RouteState",
          "method": "handleContext",
          "line": 1284
        },
        {
          "class": "io.vertx.ext.web.impl.RoutingContextImplBase",
          "method": "iterateNext",
          "line": 177
        },
        {
          "class": "io.vertx.ext.web.impl.RoutingContextImpl",
          "method": "next",
          "line": 141
        },
        {
          "class": "io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$1",
          "method": "handle",
          "line": 58
        },
        {
          "class": "io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$1",
          "method": "handle",
          "line": 36
        },
        {
          "class": "io.opentelemetry.javaagent.instrumentation.vertx.RoutingContextHandlerWrapper",
          "method": "handle",
          "line": 39
        },
        {
          "class": "io.opentelemetry.javaagent.instrumentation.vertx.RoutingContextHandlerWrapper",
          "method": "handle",
          "line": 21
        },
        {
          "class": "io.vertx.ext.web.impl.RouteState",
          "method": "handleContext",
          "line": 1284
        },
        {
          "class": "io.vertx.ext.web.impl.RoutingContextImplBase",
          "method": "iterateNext",
          "line": 177
        },
        {
          "class": "io.vertx.ext.web.impl.RoutingContextImpl",
          "method": "next",
          "line": 141
        },
        {
          "class": "org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter",
          "method": "lambda$createBlockingHandler$0",
          "line": 82
        },
        {
          "class": "io.opentelemetry.javaagent.bootstrap.executors.ContextPropagatingRunnable",
          "method": "run",
          "line": 37
        },
        {
          "class": "io.quarkus.vertx.core.runtime.VertxCoreRecorder$14",
          "method": "runWith",
          "line": 576
        },
        {
          "class": "org.jboss.threads.EnhancedQueueExecutor$Task",
          "method": "run",
          "line": 2513
        },
        {
          "class": "org.jboss.threads.EnhancedQueueExecutor$ThreadBody",
          "method": "run",
          "line": 1538
        },
        {
          "class": "org.jboss.threads.DelegatingRunnable",
          "method": "run",
          "line": 29
        },
        {
          "class": "org.jboss.threads.ThreadLocalResettingRunnable",
          "method": "run",
          "line": 29
        },
        {
          "class": "io.netty.util.concurrent.FastThreadLocalRunnable",
          "method": "run",
          "line": 30
        },
        {
          "class": "java.lang.Thread",
          "method": "run",
          "line": 833
        }
      ]
    }
  }
}
```

### How to Reproduce?

I have two client ids `source` and `target`. Both configured with client authentication required.

1. Created a new client scope named `token-exchange` and added a hardcoded claim mapper that adds it to the access token as `"token-exchange":true`.
2. Created a _client scope_ policy in the `realm-management` client, which requires a `token-exchange` scope
3. Added the scope to the list of `source` client scopes.
4.  Created a permission on the `target` client to allow a token-exchange with the policy from step 2.

Then, I request a new access token from the `source` client:
```bash
curl --location 'https://staging/realms/master/protocol/openid-connect/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'client_id=source' \
--data-urlencode 'client_secret=[REDACTED]' \
--data-urlencode 'grant_type=client_credentials' \
--data-urlencode 'scope=token-exchange'
```
And when I attempt to exchange it I get `500 Internal Server Error`.
```bash
curl --location 'https://staging/realms/master/protocol/openid-connect/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'client_id=source' \
--data-urlencode 'client_secret=[REDACTED]' \
--data-urlencode 'subject_token=TOKEN' \
--data-urlencode 'audience=target' \
--data-urlencode 'grant_type=urn:ietf:params:oauth:grant-type:token-exchange' \
--data-urlencode 'requested_token_type=urn:ietf:params:oauth:token-type:access_token'
```


### Anything else?

_No response_

@mposolda I retested on v23.0.3 and reproduced the issue again.

Here's a token-exchange request:
```bash
curl --location 'http://localhost:8080/realms/master/protocol/openid-connect/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'client_id=source' \
--data-urlencode 'client_secret=F3qWgw1IuE8FQ3LqFPrMrrW7gSR0WDQS' \
--data-urlencode 'subject_token=eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJFYjdmd09UTm84aEQ2TGtmMTFQNzJDOTEzQV9KZldndlo4bVZoWVpSd0pRIn0.eyJleHAiOjE3MDY2NDY2MDcsImlhdCI6MTcwNjY0NjU0NywianRpIjoiODg1MzUzYzUtZGJiOS00OTQ0LTg2MDUtMjRmMDk3NDcxN2U4IiwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo4MDgwL3JlYWxtcy9tYXN0ZXIiLCJhdWQiOiJhY2NvdW50Iiwic3ViIjoiM2Q1N2MxYTMtNDgzMS00MGMyLWJjMmEtZTlmNjY2NjBiMGNjIiwidHlwIjoiQmVhcmVyIiwiYXpwIjoic291cmNlIiwiYWNyIjoiMSIsImFsbG93ZWQtb3JpZ2lucyI6WyIvKiJdLCJyZWFsbV9hY2Nlc3MiOnsicm9sZXMiOlsiZGVmYXVsdC1yb2xlcy1tYXN0ZXIiLCJvZmZsaW5lX2FjY2VzcyIsInVtYV9hdXRob3JpemF0aW9uIl19LCJyZXNvdXJjZV9hY2Nlc3MiOnsiYWNjb3VudCI6eyJyb2xlcyI6WyJtYW5hZ2UtYWNjb3VudCIsIm1hbmFnZS1hY2NvdW50LWxpbmtzIiwidmlldy1wcm9maWxlIl19fSwic2NvcGUiOiJ0ZSBlbWFpbCBwcm9maWxlIiwidGUiOnRydWUsImNsaWVudEhvc3QiOiIxNzIuMTcuMC4xIiwiZW1haWxfdmVyaWZpZWQiOmZhbHNlLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJzZXJ2aWNlLWFjY291bnQtc291cmNlIiwiY2xpZW50QWRkcmVzcyI6IjE3Mi4xNy4wLjEiLCJjbGllbnRfaWQiOiJzb3VyY2UifQ.xDmkcsNsb9arkfMLXOnax4K4CYwqoV9CDBSdMSYuKJubrb39i0RN3y_nK-vKgALJomL8yuM2rMsQlXbVCigamvYoiRfYsV61IN2RFXMgzdkmiF3Z-9FlCpetsaE5GqiNbzqBhRct-Qo811z2qhgNTIK_HWQqn8RZJJCb9Kl-Um17kRW5VpYu8dPy0lkxa37YjY02U2EbkHXYuh-sAc_Vd-FkPlgkAUkzeQZxjs9MyDnLy7sFLIP8CZrWzUx6qwfIHjE67oWXNR8KQVQnsO10fo7kyjo0l10NQp4gnel1cy5g5DJrD1OVu2kQE-GAem-uQACbBEhbzuIp4kY9QePkvg' \
--data-urlencode 'audience=target' \
--data-urlencode 'grant_type=urn:ietf:params:oauth:grant-type:token-exchange' \
--data-urlencode 'requested_token_type=urn:ietf:params:oauth:token-type:access_token'
```

If I configure the `target` client id permission to allow token exchange based on client id policy, the above request succeeds.

But if I configure based on client scope policy, then the above request fails with:
```
2024-01-30 20:27:55,698 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-27) Uncaught server error: java.lang.RuntimeException: Failed to evaluate permissions
	at org.keycloak.authorization.policy.evaluation.DecisionPermissionCollector.onError(DecisionPermissionCollector.java:182)
	at org.keycloak.authorization.permission.evaluator.IterablePermissionEvaluator.evaluate(IterablePermissionEvaluator.java:71)
	at org.keycloak.authorization.permission.evaluator.IterablePermissionEvaluator.evaluate(IterablePermissionEvaluator.java:87)
	at org.keycloak.services.resources.admin.permissions.MgmtPermissions.evaluatePermission(MgmtPermissions.java:349)
	at org.keycloak.services.resources.admin.permissions.MgmtPermissions.evaluatePermission(MgmtPermissions.java:342)
	at org.keycloak.services.resources.admin.permissions.ClientPermissions.canExchangeTo(ClientPermissions.java:363)
	at org.keycloak.protocol.oidc.DefaultTokenExchangeProvider.exchangeClientToClient(DefaultTokenExchangeProvider.java:332)
	at org.keycloak.protocol.oidc.DefaultTokenExchangeProvider.tokenExchange(DefaultTokenExchangeProvider.java:242)
	at org.keycloak.protocol.oidc.DefaultTokenExchangeProvider.exchange(DefaultTokenExchangeProvider.java:128)
	at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.tokenExchange(TokenEndpoint.java:884)
	at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.processGrantRequest(TokenEndpoint.java:215)
	at org.keycloak.protocol.oidc.endpoints.TokenEndpoint$quarkusrestinvoker$processGrantRequest_6408e15340992839b66447750c221d9aaa837bd7.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.NullPointerException: Cannot invoke "org.keycloak.authorization.attribute.Attributes$Entry.asString(int)" because the return value of "org.keycloak.authorization.attribute.Attributes.getValue(String)" is null
	at org.keycloak.authorization.policy.provider.clientscope.ClientScopePolicyProvider.hasClientScope(ClientScopePolicyProvider.java:75)
	at org.keycloak.authorization.policy.provider.clientscope.ClientScopePolicyProvider.evaluate(ClientScopePolicyProvider.java:61)
	at org.keycloak.authorization.policy.provider.permission.AbstractPermissionProvider.evaluate(AbstractPermissionProvider.java:56)
	at org.keycloak.authorization.policy.provider.permission.ScopePolicyProvider.evaluate(ScopePolicyProvider.java:52)
	at org.keycloak.authorization.policy.evaluation.DefaultPolicyEvaluator.lambda$createPolicyEvaluator$0(DefaultPolicyEvaluator.java:117)
	at java.base/java.util.function.Consumer.lambda$andThen$0(Consumer.java:65)
	at java.base/java.util.function.Consumer.lambda$andThen$0(Consumer.java:65)
	at org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession$PolicyCache$1.accept(StoreFactoryCacheSession.java:967)
	at org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession$PolicyCache$1.accept(StoreFactoryCacheSession.java:962)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at org.keycloak.utils.ClosingStream.forEach(ClosingStream.java:128)
	at org.keycloak.authorization.jpa.store.JPAPolicyStore.findByResource(JPAPolicyStore.java:224)
	at org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession$PolicyCache.lambda$findByResource$4(StoreFactoryCacheSession.java:962)
	at org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession$PolicyCache.cacheQuery(StoreFactoryCacheSession.java:1075)
	at org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession$PolicyCache.cacheQuery(StoreFactoryCacheSession.java:1064)
	at org.keycloak.models.cache.infinispan.authorization.StoreFactoryCacheSession$PolicyCache.findByResource(StoreFactoryCacheSession.java:960)
	at org.keycloak.authorization.AuthorizationProvider$3.findByResource(AuthorizationProvider.java:409)
	at org.keycloak.authorization.policy.evaluation.DefaultPolicyEvaluator.evaluate(DefaultPolicyEvaluator.java:71)
	at org.keycloak.authorization.permission.evaluator.IterablePermissionEvaluator.evaluate(IterablePermissionEvaluator.java:66)
	... 20 more
```

I also noticed that if I were to remove the `audience` param from the token exchange request then it also succeeds. However, audience request is a must per RFC.
```
--data-urlencode 'audience=target' 
```

--- a/services/src/main/java/org/keycloak/authorization/common/ClientModelIdentity.java
+++ b/services/src/main/java/org/keycloak/authorization/common/ClientModelIdentity.java
@@ -16,6 +16,7 @@
  */
 package org.keycloak.authorization.common;
 
+import org.keycloak.OAuth2Constants;
 import org.keycloak.authorization.attribute.Attributes;
 import org.keycloak.authorization.identity.Identity;
 import org.keycloak.common.util.MultivaluedHashMap;
@@ -24,20 +25,27 @@ import org.keycloak.models.KeycloakSession;
 import org.keycloak.models.RealmModel;
 import org.keycloak.models.RoleModel;
 import org.keycloak.models.UserModel;
+import org.keycloak.representations.AccessToken;
 
 /**
  * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
  * @version $Revision: 1 $
  */
 public class ClientModelIdentity implements Identity {
-    protected RealmModel realm;
-    protected ClientModel client;
-    protected UserModel serviceAccount;
+    protected final RealmModel realm;
+    protected final ClientModel client;
+    protected final UserModel serviceAccount;
+    protected final AccessToken token;
 
     public ClientModelIdentity(KeycloakSession session, ClientModel client) {
-        this.realm = client.getRealm();
+        this(session, client, null);
+    }
+
+    public ClientModelIdentity(KeycloakSession session, ClientModel client, AccessToken token) {
+        this.realm = session.getContext().getRealm();
         this.client = client;
         this.serviceAccount = session.users().getServiceAccount(client);
+        this.token = token;
     }
 
     @Override
@@ -49,6 +57,9 @@ public class ClientModelIdentity implements Identity {
     public Attributes getAttributes() {
         MultivaluedHashMap map = new MultivaluedHashMap<String, String>();
         if (serviceAccount != null) map.addAll(serviceAccount.getAttributes());
+        if (token != null) {
+            map.add(OAuth2Constants.SCOPE, token.getScope());
+        }
         return Attributes.from(map);
     }
 

--- a/services/src/main/java/org/keycloak/protocol/oidc/DefaultTokenExchangeProvider.java
+++ b/services/src/main/java/org/keycloak/protocol/oidc/DefaultTokenExchangeProvider.java
@@ -332,7 +332,7 @@ public class DefaultTokenExchangeProvider implements TokenExchangeProvider {
                 // public clients can not exchange tokens from other client
                 forbiddenIfClientIsNotTokenHolder(disallowOnHolderOfTokenMismatch, tokenHolder);
             }
-            if (!AdminPermissions.management(session, realm).clients().canExchangeTo(client, targetClient)) {
+            if (!AdminPermissions.management(session, realm).clients().canExchangeTo(client, targetClient, token)) {
                 event.detail(Details.REASON, "client not allowed to exchange to audience");
                 event.error(Errors.NOT_ALLOWED);
                 throw new CorsErrorResponseException(cors, OAuthErrorException.ACCESS_DENIED, "Client not allowed to exchange", Response.Status.FORBIDDEN);

--- a/services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionManagement.java
+++ b/services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionManagement.java
@@ -20,6 +20,7 @@ import org.keycloak.authorization.model.Policy;
 import org.keycloak.authorization.model.Resource;
 import org.keycloak.authorization.model.ResourceServer;
 import org.keycloak.models.ClientModel;
+import org.keycloak.representations.AccessToken;
 
 import java.util.Map;
 
@@ -42,6 +43,8 @@ public interface ClientPermissionManagement {
     Map<String, String> getPermissions(ClientModel client);
 
     boolean canExchangeTo(ClientModel authorizedClient, ClientModel to);
+
+    boolean canExchangeTo(ClientModel authorizedClient, ClientModel to, AccessToken token);
 
     Policy exchangeToPermission(ClientModel client);
 

--- a/services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java
+++ b/services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java
@@ -32,6 +32,7 @@ import org.keycloak.models.ClientModel;
 import org.keycloak.models.ClientScopeModel;
 import org.keycloak.models.KeycloakSession;
 import org.keycloak.models.RealmModel;
+import org.keycloak.representations.AccessToken;
 import org.keycloak.representations.idm.authorization.Permission;
 import org.keycloak.storage.StorageId;
 
@@ -319,7 +320,12 @@ class ClientPermissions implements ClientPermissionEvaluator,  ClientPermissionM
     }
 
     @Override
-    public boolean canExchangeTo(ClientModel authorizedClient, ClientModel to) {
+    public boolean canExchangeTo(ClientModel authorizedClient, ClientModel client) {
+        return canExchangeTo(authorizedClient, client, null);
+    }
+
+    @Override
+    public boolean canExchangeTo(ClientModel authorizedClient, ClientModel to, AccessToken token) {
 
         ResourceServer server = resourceServer(to);
         if (server == null) {
@@ -351,7 +357,7 @@ class ClientPermissions implements ClientPermissionEvaluator,  ClientPermissionM
             logger.debug(TOKEN_EXCHANGE + " not initialized");
             return false;
         }
-        ClientModelIdentity identity = new ClientModelIdentity(session, authorizedClient);
+        ClientModelIdentity identity = new ClientModelIdentity(session, authorizedClient, token);
         EvaluationContext context = new DefaultEvaluationContext(identity, session) {
             @Override
             public Map<String, Collection<String>> getBaseAttributes() {

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/ClientTokenExchangeTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/ClientTokenExchangeTest.java
@@ -67,7 +67,6 @@ import jakarta.ws.rs.core.Response;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
-import java.util.concurrent.TimeUnit;
 
 import static org.hamcrest.Matchers.instanceOf;
 import static org.junit.Assert.assertEquals;

